File: dart_sample/lib/demos/general/nested_scroll_view_demo_page.dart
import 'package:material_ui/material_ui.dart';

class NestedScrollViewDemoPage extends StatelessWidget {
  const NestedScrollViewDemoPage({super.key});

  @override
  Widget build(BuildContext context) {
    return NestedScrollView(
      headerSliverBuilder: (BuildContext headerContext, bool innerBoxIsScrolled) {
        return <Widget>[
          SliverOverlapAbsorber(
            handle: NestedScrollView.sliverOverlapAbsorberHandleFor(headerContext),
            sliver: SliverPersistentHeader(
              pinned: true,
              delegate: _NestedScrollViewHeaderDelegate(innerBoxIsScrolled),
            ),
          ),
          const SliverToBoxAdapter(
            child: ColoredBox(
              color: Color(0xFFE8EAF6),
              child: Padding(
                padding: EdgeInsets.symmetric(horizontal: 16, vertical: 14),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.stretch,
                  spacing: 4,
                  children: <Widget>[
                    Text(
                      'Outer header sliver',
                      style: TextStyle(fontSize: 18, color: Colors.black),
                    ),
                    Text(
                      'This scrolls away completely before the body starts scrolling.',
                      style: TextStyle(fontSize: 13, color: Colors.black54),
                    ),
                  ],
                ),
              ),
            ),
          ),
        ];
      },
      body: Builder(
        builder: (BuildContext bodyContext) {
          return CustomScrollView(
            slivers: <Widget>[
              SliverOverlapInjector(
                handle: NestedScrollView.sliverOverlapAbsorberHandleFor(bodyContext),
              ),
              SliverFixedExtentList.builder(
                itemCount: 40,
                itemExtent: 46,
                itemBuilder: (BuildContext context, int index) {
                  return Container(
                    color: index.isEven
                        ? const Color(0xFFF5F5F5)
                        : const Color(0xFFFFFFFF),
                    padding: const EdgeInsets.symmetric(
                      horizontal: 16,
                      vertical: 12,
                    ),
                    child: Text(
                      'body row #$index',
                      style: const TextStyle(fontSize: 14, color: Colors.black),
                    ),
                  );
                },
              ),
            ],
          );
        },
      ),
    );
  }
}

class _NestedScrollViewHeaderDelegate extends SliverPersistentHeaderDelegate {
  _NestedScrollViewHeaderDelegate(this.innerBoxIsScrolled);

  final bool innerBoxIsScrolled;

  @override
  double get minExtent => 72;

  @override
  double get maxExtent => 72;

  @override
  Widget build(BuildContext context, double shrinkOffset, bool overlapsContent) {
    return SizedBox(
      height: 72,
      child: ColoredBox(
        color: innerBoxIsScrolled
            ? const Color(0xFF90CAF9)
            : const Color(0xFFBBDEFB),
        child: Padding(
          padding: const EdgeInsets.symmetric(horizontal: 16, vertical: 12),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            crossAxisAlignment: CrossAxisAlignment.stretch,
            spacing: 2,
            children: <Widget>[
              const Text(
                'NestedScrollView',
                style: TextStyle(fontSize: 20, color: Colors.black),
              ),
              Text(
                innerBoxIsScrolled
                    ? 'innerBoxIsScrolled: true'
                    : 'innerBoxIsScrolled: false',
                style: const TextStyle(fontSize: 13, color: Colors.black54),
              ),
            ],
          ),
        ),
      ),
    );
  }

  @override
  bool shouldRebuild(_NestedScrollViewHeaderDelegate oldDelegate) {
    return oldDelegate.innerBoxIsScrolled != innerBoxIsScrolled;
  }
}

File: src/Sample/Plumix.Sample/Demos/General/NestedScrollViewDemoPage.cs
using Avalonia;
using Avalonia.Media;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/general/nested_scroll_view_demo_page.dart (exact sample parity)

namespace Plumix;

public sealed class NestedScrollViewDemoPage : StatelessWidget
{
    public override Widget Build(BuildContext context)
    {
        return new NestedScrollView(
            headerSliverBuilder: (headerContext, innerBoxIsScrolled) =>
            [
                new SliverOverlapAbsorber(
                    handle: NestedScrollView.SliverOverlapAbsorberHandleFor(headerContext),
                    sliver: new SliverPersistentHeader(
                        pinned: true,
                        @delegate: new NestedScrollViewHeaderDelegate(innerBoxIsScrolled))),
                new SliverToBoxAdapter(
                    new Container(
                        color: Color.Parse("#FFE8EAF6"),
                        padding: new Thickness(16, 14),
                        child: new Column(
                            crossAxisAlignment: CrossAxisAlignment.Stretch,
                            spacing: 4,
                            children:
                            [
                                new Text("Outer header sliver", fontSize: 18, color: Colors.Black),
                                new Text(
                                    "This scrolls away completely before the body starts scrolling.",
                                    fontSize: 13,
                                    color: Colors.DimGray),
                            ]))),
            ],
            body: new Builder(bodyContext => new CustomScrollView(
            [
                new SliverOverlapInjector(
                    NestedScrollView.SliverOverlapAbsorberHandleFor(bodyContext)),
                SliverFixedExtentList.Builder(
                    itemCount: 40,
                    itemExtent: 46,
                    itemBuilder: (_, index) => new Container(
                        color: index % 2 == 0
                            ? Color.Parse("#FFF5F5F5")
                            : Color.Parse("#FFFFFFFF"),
                        padding: new Thickness(16, 12),
                        child: new Text($"body row #{index}", fontSize: 14, color: Colors.Black)),
                    addAutomaticKeepAlives: false),
            ])));
    }
}

/// <summary>The pinned header of the nested scroll view demo, which reacts to the body scrolling.</summary>
internal sealed class NestedScrollViewHeaderDelegate(bool innerBoxIsScrolled) : SliverPersistentHeaderDelegate
{
    public override double MinExtent => 72;

    public override double MaxExtent => 72;

    public override Widget Build(BuildContext context, double shrinkOffset, bool overlapsContent)
    {
        return new Container(
            height: 72,
            color: innerBoxIsScrolled
                ? Color.Parse("#FF90CAF9")
                : Color.Parse("#FFBBDEFB"),
            padding: new Thickness(16, 12),
            child: new Column(
                mainAxisAlignment: MainAxisAlignment.Center,
                crossAxisAlignment: CrossAxisAlignment.Stretch,
                spacing: 2,
                children:
                [
                    new Text("NestedScrollView", fontSize: 20, color: Colors.Black),
                    new Text(
                        innerBoxIsScrolled ? "innerBoxIsScrolled: true" : "innerBoxIsScrolled: false",
                        fontSize: 13,
                        color: Colors.DimGray),
                ]));
    }

    public override bool ShouldRebuild(SliverPersistentHeaderDelegate oldDelegate)
    {
        return oldDelegate is not NestedScrollViewHeaderDelegate other
               || other.InnerBoxIsScrolled != innerBoxIsScrolled;
    }

    private bool InnerBoxIsScrolled => innerBoxIsScrolled;
}
